plugins/test/framework.cc

```diff
@@ -20,13 +20,54 @@
 
 namespace service_extensions_samples {
 
+size_t Buffer::size() const {
+  if (owned_string_buffer_) {
+    return owned_string_buffer_->length();
+  }
+  return proxy_wasm::BufferBase::size();
+}
+proxy_wasm::WasmResult Buffer::copyTo(proxy_wasm::WasmBase* wasm, size_t start,
+                                      size_t length, uint64_t ptr_ptr,
+                                      uint64_t size_ptr) const {
+  if (owned_string_buffer_) {
+    auto buffer_section = owned_string_buffer_->substr(start, length);
+    std::string_view s(buffer_section);
+    if (!wasm->copyToPointerSize(s, ptr_ptr, size_ptr)) {
+      return proxy_wasm::WasmResult::InvalidMemoryAccess;
+    }
+    return proxy_wasm::WasmResult::Ok;
+  }
+  return proxy_wasm::BufferBase::copyTo(wasm, start, length, ptr_ptr, size_ptr);
+}
+
+proxy_wasm::WasmResult Buffer::copyFrom(size_t start, size_t length,
+                                        std::string_view data) {
+  if (!owned_string_buffer_) {
+    owned_string_buffer_ = std::string(data_);
+  }
+  if (start == 0) {
```

Comment:
I think this can be simplified significantly if you use `std::string::replace()`, which does pretty much everything you need. i.e. something like:

```
owned_string_buffer_.replace(start, length, data);
```